src/coreclr/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2125,38 +2127,119 @@ void CodeGen::genFMAIntrinsic(GenTreeHWIntrinsic* node)
     regNumber op1Reg;
     regNumber op2Reg;
 
+    regNumber op1RegOld = op1->GetRegNum();
+    regNumber op2RegOld = op2->GetRegNum();
+    regNumber op3RegOld = op3->GetRegNum();
+
+    // For special case that targetReg equals to more than one op Regs
+    bool sameReg = false;
+    if ((op1RegOld == op2RegOld && targetReg == op1RegOld) || (op1RegOld == op3RegOld && targetReg == op1RegOld) ||
+        (op2RegOld == op3RegOld && targetReg == op1RegOld))
+        sameReg = true;
+
     bool       isCommutative   = false;
     const bool copiesUpperBits = HWIntrinsicInfo::CopiesUpperBits(intrinsicId);
 
     // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
     assert(!copiesUpperBits || !op1->isContained());
 
-    if (op2->isContained() || op2->isUsedFromSpillTemp())
+    // Intrinsics with CopyUpperBits semantics must have op1 as target
+    if ((targetReg == op1RegOld || copiesUpperBits) && !sameReg)
     {
-        // 132 form: op1 = (op1 * op3) + [op2]
-
-        ins    = (instruction)(ins - 1);
-        op1Reg = op1->GetRegNum();
-        op2Reg = op3->GetRegNum();
-        op3    = op2;
+        if (op2->isContained() || op2->IsRegOptional())
+        {
+            // op1 = (op1 * [op2]) + op3
+            // 132 form: XMM1 = (XMM1 * [XMM3]) + XMM2
+            ins    = _132form;
+            op1Reg = op1RegOld;
+            op2Reg = op3RegOld;
+            op3    = op2;
+        }
+        else
+        {
+            assert(op3->isContained() || op3->IsRegOptional());
+            // op1 = (op1 * op2) + [op3]
+            // 213 form: XMM1 = (XMM2 * XMM1) + [XMM3]
+            op1Reg        = op1RegOld;
```

Comment:
I would just have `op1Reg = op1RegOld` once at the top before the `if (op2->isContained() || op2->IsRegOptional())`.